File: ios/Runner/Activity/AnnouncementStateCache.swift
import Foundation

/// Keeps live announcement progress in memory between durable checkpoints.
final class AnnouncementStateCache {
    static let defaultCheckpointInterval: TimeInterval = 60

    private let checkpointInterval: TimeInterval
    private var persistedState: AnnouncementStateData?
    private var lastCheckpointUptime: TimeInterval?

    private(set) var state: AnnouncementStateData?

    init(checkpointInterval: TimeInterval = defaultCheckpointInterval) {
        precondition(checkpointInterval > 0)
        self.checkpointInterval = checkpointInterval
    }

    func restore(_ state: AnnouncementStateData, uptime: TimeInterval) {
        self.state = state
        persistedState = state
        lastCheckpointUptime = uptime
    }

    func update(_ state: AnnouncementStateData) {
        self.state = state
    }

    func stateToPersist(
        uptime: TimeInterval,
        force: Bool = false
    ) -> AnnouncementStateData? {
        guard let state, state != persistedState else {
            return nil
        }
        guard let lastCheckpointUptime else {
            return state
        }
        return force || uptime - lastCheckpointUptime >= checkpointInterval
            ? state
            : nil
    }

    func markPersisted(uptime: TimeInterval) {
        persistedState = state
        lastCheckpointUptime = uptime
    }

    func reset() {
        state = nil
        persistedState = nil
        lastCheckpointUptime = nil
    }
}

File: android/app/src/main/kotlin/com/endurain/endurain/activity/AnnouncementStateCache.kt
package com.endurain.endurain.activity

/** Keeps live announcement progress in memory between durable checkpoints. */
internal class AnnouncementStateCache(
    private val checkpointIntervalMillis: Long = DEFAULT_CHECKPOINT_INTERVAL_MILLIS,
) {
    private var persistedState: AnnouncementStateData? = null
    private var lastCheckpointUptimeMillis: Long? = null

    var state: AnnouncementStateData? = null
        private set

    init {
        require(checkpointIntervalMillis > 0)
    }

    fun restore(state: AnnouncementStateData, uptimeMillis: Long) {
        this.state = state
        persistedState = state
        lastCheckpointUptimeMillis = uptimeMillis
    }

    fun update(state: AnnouncementStateData) {
        this.state = state
    }

    fun stateToPersist(
        uptimeMillis: Long,
        force: Boolean = false,
    ): AnnouncementStateData? {
        val current = state ?: return null
        if (current == persistedState) {
            return null
        }
        val lastCheckpoint = lastCheckpointUptimeMillis ?: return current
        return if (force || uptimeMillis - lastCheckpoint >= checkpointIntervalMillis) {
            current
        } else {
            null
        }
    }

    fun markPersisted(uptimeMillis: Long) {
        persistedState = state
        lastCheckpointUptimeMillis = uptimeMillis
    }

    fun reset() {
        state = null
        persistedState = null
        lastCheckpointUptimeMillis = null
    }

    companion object {
        const val DEFAULT_CHECKPOINT_INTERVAL_MILLIS = 60_000L
    }
}